### PR TITLE
Add explain analyze detailed info of hash agg

### PR DESCRIPTION
**Pull request by yaowangm**
### Summary

After upgrading to GPDB7, we lost the detailed cdb executor instruments of hashagg since the code base has been changed in a large size. The patch is to re-implement explain analyze related code of hashagg to provide more critical info for troubleshooting issues.

### Example

We use a script to create the hashagg scenario:
```
create table streaming_t1 (d int, i int, c int, dt date);
insert into streaming_t1 select i/10, i/10, i/100, '2009-06-10'::date + ( (i%34) || ' days')::interval from generate_series(0, 99999999) i;

analyze streaming_t1;

SET statement_mem='1000kB';
SET optimizer=off;

explain analyze select count(distinct i) from streaming_t1 group by c;
```

Result in GPDB6:

```
testdb=# explain analyze select count(distinct i) from streaming_t1 group by c;
                                                                                 QUERY PLAN                                                      
                           
-------------------------------------------------------------------------------------------------------------------------------------------------
---------------------------
 Gather Motion 3:1  (slice2; segments: 3)  (cost=3117970.90..3122369.13 rows=439823 width=12) (actual time=15782.991..16337.777 rows=1000000 loop
s=1)
   ->  HashAggregate  (cost=3117970.90..3122369.13 rows=146608 width=12) (actual time=15785.477..16234.622 rows=334042 loops=1)
         Group Key: streaming_t1.c
         Extra Text: (seg2)   334042 groups total in 32 batches; 1 overflows; 2999062 spill groups.
 (seg2)   Hash chain length 1.5 avg, 11 max, using 277170 of 557056 buckets; total 1 expansions.
 
         ->  HashAggregate  (cost=2838677.18..2967972.60 rows=3333296 width=8) (actual time=13428.779..14403.499 rows=3340420 loops=1)
               Group Key: streaming_t1.c, streaming_t1.i
               Extra Text: (seg2)   3340420 groups total in 32 batches; 1 overflows; 3340420 spill groups.
 (seg2)   Hash chain length 3.3 avg, 16 max, using 1036832 of 1081344 buckets; total 1 expansions.
 
               ->  Redistribute Motion 3:3  (slice1; segments: 3)  (cost=1884391.12..2634382.62 rows=3333296 width=8) (actual time=320.177..11683
.206 rows=3340420 loops=1)
                     Hash Key: streaming_t1.c
                     ->  HashAggregate  (cost=1884391.12..2434384.88 rows=3333296 width=8) (actual time=1074.721..11839.608 rows=3334839 loops=1)
                           Group Key: streaming_t1.c, streaming_t1.i
                           Extra Text: (seg2)   Hash chain length 3.5 avg, 15 max, using 949145 of 983040 buckets; total 1 expansions.
 
                           ->  Seq Scan on streaming_t1  (cost=0.00..1134399.64 rows=33332955 width=8) (actual time=1.521..6235.343 rows=33348390
 loops=1)
 Planning time: 0.380 ms
   (slice0)    Executor memory: 151K bytes.
   (slice1)    Executor memory: 5031K bytes avg x 3 workers, 5031K bytes max (seg0).
 * (slice2)    Executor memory: 10905K bytes avg x 3 workers, 10905K bytes max (seg0).  Work_mem: 5001K bytes max, 132225K bytes wanted.
 Memory used:  10000kB
 Memory wanted:  396972kB
 Optimizer: Postgres query optimizer
 Execution time: 16406.703 ms
(26 rows) 
```

Result in GPDB7 without the patch:
```
testdb=# explain analyze select count(distinct i) from streaming_t1 group by c;
                                                                               QUERY PLAN--------------------------------------------------------------------------------------------------------
-----------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=1257397.71..1264900.21 rows=450150 width=12) (actual ti
me=17254.426..19779.878 rows=1000000 loops=1)
   ->  HashAggregate  (cost=1257397.71..1258898.21 rows=150050 width=12) (actual time=17273.667..19759.6
68 rows=334042 loops=1)
         Group Key: c
         Planned Partitions: 4
         Peak Memory Usage: 0 kB
         ->  HashAggregate  (cost=1197800.36..1240730.74 rows=3333393 width=8) (actual time=12139.638..1
6350.869 rows=3340420 loops=1)
               Group Key: c, i
               Planned Partitions: 4
               Peak Memory Usage: 0 kB
               ->  Redistribute Motion 3:3  (slice2; segments: 3)  (cost=758358.40..1085713.88 rows=9826
761 width=8) (actual time=8.817..11453.585 rows=3341861 loops=1)
                     Hash Key: c
                     ->  Streaming HashAggregate  (cost=758358.40..889178.67 rows=9826761 width=8) (actu
al time=7.587..11070.286 rows=3336270 loops=1)
                           Group Key: c, i
                           Planned Partitions: 4
                           Peak Memory Usage: 0 kB
                           ->  Seq Scan on streaming_t1  (cost=0.00..378143.28 rows=33333928 width=8) (a
ctual time=0.074..3261.548 rows=33348390 loops=1)
 Optimizer: Postgres query optimizer
 Planning Time: 1.224 ms
   (slice0)    Executor memory: 349K bytes.
   (slice1)    Executor memory: 1181K bytes avg x 3 workers, 1181K bytes max (seg0).  Work_mem: 785K byt
es max.
   (slice2)    Executor memory: 198K bytes avg x 3 workers, 199K bytes max (seg1).  Work_mem: 369K bytes
 max.
 Memory used:  1000kB
 Execution Time: 19873.473 ms
(23 rows) 
```
Result in GPDB7 with the patch:
```
testdb=# explain analyze select count(distinct i) from streaming_t1 group by c;
                                                                                                           QUERY PLA
N
--------------------------------------------------------------------------------------------------------------------
-------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=1257397.71..1264859.84 rows=447728 width=12) (actual time=37990.820
..43252.288 rows=1000000 loops=1)
   ->  HashAggregate  (cost=1257397.71..1258890.13 rows=149243 width=12) (actual time=38041.124..43104.985 rows=3340
42 loops=1)
         Group Key: c
         Planned Partitions: 4
         Extra Text: (seg2)   hash table(s): 1; 334042 groups total in 340 batches, 13253020 spill partitions; disk
usage: 111584KB; chain length 2.6 avg, 13 max; using 334042 of 1396736 buckets; total 0 expansions.

         ->  HashAggregate  (cost=1197800.36..1240730.74 rows=3333393 width=8) (actual time=27682.691..36143.712 row
s=3340420 loops=1)
               Group Key: c, i
               Planned Partitions: 4
               Extra Text: (seg2)   hash table(s): 1; 3340420 groups total in 1364 batches, 13359480 spill partition
s; disk usage: 112128KB; chain length 2.7 avg, 16 max; using 3340420 of 5591040 buckets; total 0 expansions.

               ->  Redistribute Motion 3:3  (slice2; segments: 3)  (cost=758358.40..1085713.88 rows=9826761 width=8)
 (actual time=20.442..26186.948 rows=3341861 loops=1)
                     Hash Key: c
                     ->  Streaming HashAggregate  (cost=758358.40..889178.67 rows=9826761 width=8) (actual time=17.5
70..25616.621 rows=3336270 loops=1)
                           Group Key: c, i
                           Planned Partitions: 4
                           Extra Text: (seg2)   hash table(s): 1; chain length 2.6 avg, 15 max; using 3336270 of 586
5472 buckets; total 0 expansions.

                           ->  Seq Scan on streaming_t1  (cost=0.00..378143.28 rows=33333928 width=8) (actual time=0
.098..6420.498 rows=33348390 loops=1)
 Optimizer: Postgres query optimizer
 Planning Time: 0.717 ms
   (slice0)    Executor memory: 349K bytes.
 * (slice1)    Executor memory: 1181K bytes avg x 3 workers, 1181K bytes max (seg0).  Work_mem: 785K bytes max, 3379
05K bytes wanted.
 * (slice2)    Executor memory: 198K bytes avg x 3 workers, 199K bytes max (seg1).  Work_mem: 369K bytes max, 526874
K bytes wanted.
 Memory used:  1000kB
 Memory wanted:  1580920kB
 Execution Time: 43408.876 ms
(27 rows)
```
In a brief, the patch updated some info which are different between GPDB6 and GPDB7 (without the patch):

1. Added extra lines of "Extra Text" about more details.
2. Added extra lines of "Memory wanted".
3. Removed the info of "Peak Memory Usage" of GPDB7 (which was always 0 without the patch) since we can get it in "Executor memory". See code in `hash_agg_update_metrics()`:
```
2002         instrument->workmemused = aggstate->hash_mem_peak;
```
### Details in "Extra Text"
1. On GPDB6, there is only one hash table. But on GPDB7, because of group set, there might be multiple hash tables. We must collect statistic info of all of them. 
2. The hash table implementation was rewritten completely on GPDB7, so some concepts are meaningless and were removed. e.g. `num_overflows` is used to count `SpillSet` ever created on GPDB6, but there is no `SpillSet` on GPDB7 at all.
3. The behavior of hash table was changed greatly. On GPDB6, when there is no space for a new entry, the hash table will be spilled to disk to make room (see `spill_hash_table()`). On GPDB7, a new group will be spilled to disk in spill mode, and the hash table itself will not (see `lookup_hash_entry()`). Thus, `num_spill_groups` which indicates number of spilled groups in `spill_hash_table()` is meaningless and removed.
4. To measure the spilled data to replace `num_spill_groups`, on GPDB7 I used `num_spill_parts` to indicate the number of spilled partitions. When entering spill mode, a number of partitions will be allocated and assigned to tape set, which measures the spilled data in some degree (see `hashagg_spill_init()`). Not very sure if it's proper. Or we can still count the spilled branches on GPDB7?
5. Some concepts with same names have different meanings. e.g. on GPDB6 `num_batches` means "number of batch files" on GPDB6 (see `spill_hash_table()`), but on GPDB7 batch means "work to be done for one pass of hash aggregation (with only one grouping set)" (see definition of `HashAggBatch`). So we have "batches" output in "Extra Text" on both versions, but they have different meanings.
6. The calculation of hash chain is totally different. Please see details in `SH_COLL_STAT`().

### Memory wanted

On GPDB6, the calculation of "Memory wanted" is based on hash table implementation and not applicable on GPDB7. I used a different way (with similar theory). Please see `hash_agg_update_metrics().`

### Test

The impact for test cases is smaller than estimated because the extra explain info is available when ANALYZE is enabled. Actually all existing test cases are not impacted, and I just need a test case to cover the scenario. A candidate is src/test/regress/sql/workfile/hashagg_spill.sql, but it focuses on work file status under various spilling cases and not so good for the scenarios. So I chose src/test/isolation2/sql/spilling_hashagg.sql as a simple example, and enlarged the test data to create spilling scenario.

### CODE REVIEW

File: src/backend/executor/nodeAgg.c [hunk 1]
```diff
@@ -1982,6 +2000,45 @@ hash_agg_update_metrics(AggState *aggstate, bool from_tape, int npartitions)
...
+
+		/* Scan all perhashs and collect hash table statistic info */
+		for (int setno = 0; setno < aggstate->num_hashes; setno++)
+		{
+			AggStatePerHash perhash = &aggstate->perhash[setno];
+			tuplehash_hash *hashtab = perhash->hashtable->hashtab;
+
+			Assert(hashtab);
+
+			perhash->num_expansions += hashtab->num_expansions;
```
Comment (adam8157) [inline src/backend/executor/nodeAgg.c:2025]: Why it's `+=` here?

File: src/backend/executor/nodeAgg.c [hunk 2]
```diff
@@ -5109,3 +5177,75 @@ ReuseHashTable(AggState *node)
...
 			!bms_overlap(node->ss.ps.chgParam, aggnode->aggParams));
 }
+
+/*
+ * Append hashtable statistics to explain buffer for EXPLAIN ANALYZE
+ */
+void
+agg_hash_explain(AggState *aggstate)
+{
+	StringInfo hbuf = aggstate->ss.ps.cdbexplainbuf;
```
Comment (adam8157) [inline src/backend/executor/nodeAgg.c:5187]: Greenplum only displays one node's `cdbexplainbuf`, does it?

File: src/backend/executor/nodeAgg.c [hunk 3]
```diff
@@ -5109,3 +5177,75 @@ ReuseHashTable(AggState *node)
...
+		if (aggstate->hash_ever_spilled)
+		{
+			sum_output_groups += perhash->num_output_groups;
+			sum_spill_parts += perhash->num_spill_parts;
+		}
+
+		/* inner hash table statistic info */
+		if (perhash->chain_count > 0)
+		{
+			sum_chain_length_total += perhash->chain_length_total;
```
Comment (adam8157) [inline src/backend/executor/nodeAgg.c:5218]: How does postgres deal with statics for grouping sets? Also add them up?

File: src/backend/commands/explain.c [hunk 4]
```diff
@@ -3624,7 +3624,17 @@ show_hashagg_info(AggState *aggstate, ExplainState *es)
...
+	 * CDB:
+	 * The output code of EXPLAIN ANALYZE is from upstream, and the info is
+	 * collected in hash_agg_update_metrics(). On GPDB,
+	 * hash_agg_update_metrics() was never called on QD, and no output here.
+	 * The info is collected on QE in a different way, outputted in
+	 * agg_hash_explain(), and avaialble in "Extra Text" of EXPLAIN ANALYZE
+	 * output.
+	 * So, just return unconditionally here.
+	 */
+	//if (!es->analyze)
```
Comment (adam8157) [inline src/backend/commands/explain.c:3637]: If we stay with the current implementation of this PR, suggest checking one key variable here instead, in case utility/coordinator-only mode or some corner plans having the agg running on the coornidator? Not sure if this makes sense.

File: src/test/isolation2/sql/spilling_hashagg.sql [hunk 5]
```diff
@@ -10,9 +10,9 @@ DROP TABLE IF EXISTS test_hashagg_off;
 
 CREATE TABLE test_src_tbl AS
 WITH cte1 AS (
-    SELECT field5 from generate_series(1,1000) field5
+    SELECT field5 from generate_series(1, 3000000) field5
```
Comment (adam8157) [inline src/test/isolation2/sql/spilling_hashagg.sql:13]: This likely will take some time, maybe too long for CI.

File: src/backend/executor/nodeAgg.c [hunk 6]
```diff
@@ -5109,3 +5177,75 @@ ReuseHashTable(AggState *node)
 			!node->streaming &&
 			!bms_overlap(node->ss.ps.chgParam, aggnode->aggParams));
 }
+
+/*
+ * Append hashtable statistics to explain buffer for EXPLAIN ANALYZE
+ */
+void
+agg_hash_explain(AggState *aggstate)
```
Comment (adam8157) [inline src/backend/executor/nodeAgg.c:5185]: We probably could learn how to display CDB details from `show_sort_info` and `show_hash_info` on 6X (7X they use `shared_info`).

**Review by adam8157 — COMMENTED** [review]
```
/*
 * GPDB_12_MERGE_FIXME: we lost the detailed cdb executor instruments to print
 * by explain, which were in execHHashagg.c
 */
```

This FIXME could be removed.

**Comment by yaowangm** [comment]
> ```
> /*
>  * GPDB_12_MERGE_FIXME: we lost the detailed cdb executor instruments to print
>  * by explain, which were in execHHashagg.c
>  */
> ```
> 
> This FIXME could be removed.

Thanks for pointing it out. A real stupid cursoriness.

**Comment by yaowangm** [comment]
Some analysis of **grouping sets**: 

The behaviors of grouping sets are more complex than estimated. The behaviors under GPDB6/7 and planner/ORCA are different. I have updated `spilling_hashagg.sql` to cover the scenario of grouping sets, so we can talk based on the test case.

The query sql:
```
EXPLAIN (analyze, costs off) SELECT a, avg(b) AS b FROM test_src_tbl GROUP BY grouping sets ((a), (b));
```
### On GPDB7
Under planner, the plan works as expected: beyond Seq Scan, there is a Partial HashAggregate node to perform the hash based on hash keys a **and** b, and in "Extra Text" we can find "hash table(s): 2".
```
               ->  Partial HashAggregate (actual time=185.611..625.432 rows=66924 loops=1)
                     Hash Key: a
                     Hash Key: b
                     Extra Text: (seg1)   hash table(s): 2; 66656 groups total in 168 batches, 1319068 spill partitions; disk usage: 15104KB; chain length 2.1 avg, 6 max; using 33328 of 346112 buckets; total 0 expansions.
``` 
(See the complete plan in src/test/isolation2/expected/spilling_hashagg.out)

But under ORCA, the behavior is total different. The plan includes two Shared Scan nodes on the same table, and two separated hashagg nodes in top of them with hash key a **or** b. So it seems ORCA performed two hashagg **with two separated hashagg nodes** based on shared scan. I didn't find a hashagg node with two hash tables.
```
         ->  Append (actual time=322.666..1034.556 rows=66924 loops=1)
               ->  Finalize HashAggregate (actual time=322.665..633.635 rows=33462 loops=1)
                     Group Key: share0_ref2.b
                     Extra Text: (seg2)   hash table(s): 1; 33211 groups total in 338 batches, 657460 spill partitions; disk usage: 15936KB; chain length 1.9 avg, 5 max; using 33211 of 694272 buckets; total 0 expansions.

                     ->  Redistribute Motion 3:3  (slice2; segments: 3) (actual time=0.064..245.545 rows=167310 loops=1)
                           Hash Key: share0_ref2.b
                           ->  Streaming Partial HashAggregate (actual time=90.762..377.627 rows=167310 loops=1)
                                 Group Key: share0_ref2.b
                                 Extra Text: (seg0)   hash table(s): 1; chain length 2.0 avg, 5 max; using 167310 of 966656 buckets; total 0 expansions.
                                 ->  Shared Scan (share slice:id 2:0) (actual time=90.345..141.629 rows=167310 loops=1)
               ->  HashAggregate (actual time=119.250..393.172 rows=33462 loops=1)
                     Group Key: share0_ref3.a
                     Extra Text: (seg0)   hash table(s): 1; 33462 groups total in 340 batches, 662480 spill partitions; disk usage: 9504KB; chain length 1.9 avg, 5 max; using 33462 of 698368 buckets; total 0 expansions.
                     ->  Shared Scan (share slice:id 1:0) (actual time=0.012..48.720 rows=167310 loops=1)
```
Based on a number of tryings I never see ORCA generates plan for grouping sets like planner. Maybe we can talked to ORCA to confirm that. Anyway, the explain code works well.

### On GPDB6
With the same query and a bit more memory limit:
```
testdb=# set optimizer_enable_hashagg=on;
testdb=# SET statement_mem='3000kB';
```
ORCA generated plan like this (similar to GPDB7):
```
 Gather Motion 3:1  (slice2; segments: 3) (actual time=248.828..802.326 rows=200000 loops=1)
   ->  Sequence (actual time=248.251..785.303 rows=66924 loops=1)
         ->  Shared Scan (share slice:id 2:0) (actual time=35.343..42.686 rows=167310 loops=1)
               ->  Materialize (never executed)
                     ->  Seq Scan on test_src_tbl (actual time=0.012..17.864 rows=167310 loops=1)
         ->  Append (actual time=169.032..701.380 rows=66924 loops=1)
               ->  Result (actual time=169.032..219.725 rows=33462 loops=1)
                     ->  HashAggregate (actual time=169.030..216.492 rows=33462 loops=1)
                           Group Key: share0_ref2.b
                           Extra Text: (seg0)   33462 groups total in 32 batches; 1 overflows; 167310 spill groups.
 (seg0)   Hash chain length 1.5 avg, 10 max, using 27033 of 59392 buckets; total 6 expansions.
 
                           ->  Redistribute Motion 3:3  (slice1; segments: 3) (actual time=0.104..47.462 rows=167310 loops=1)
                                 Hash Key: share0_ref2.b
                                 ->  Result (actual time=0.035..29.166 rows=167310 loops=1)
                                       ->  Shared Scan (share slice:id 1:0) (actual time=0.034..9.259 rows=167310 loops=1)
               ->  Result (actual time=215.256..263.936 rows=33462 loops=1)
                     ->  HashAggregate (actual time=215.254..261.072 rows=33462 loops=1)
                           Group Key: share0_ref3.a
                           Extra Text: (seg0)   33462 groups total in 32 batches; 1 overflows; 167310 spill groups.
 (seg0)   Hash chain length 1.5 avg, 9 max, using 27009 of 59392 buckets; total 6 expansions.
 
                           ->  Shared Scan (share slice:id 2:0) (actual time=0.020..7.671 rows=167310 loops=1)
```
Planner need even more memory limit (4000kb) and generated a different plan. Basically, it also used two Shared Scan nodes, and separated groupagg/hashagg in top of them. The strategy is similar to ORCA with additional groupagg nodes. The important thing is, **planner didn't generate a hashagg node with two hash keys in two hash tables**. It makes sense because GPDB6 code doesn't support multiple hash tables in a hashagg node. See the plan:
```
 Gather Motion 3:1  (slice3; segments: 3) (actual time=601.177..850.561 rows=200000 loops=1)
   ->  Append (actual time=600.452..811.547 rows=66777 loops=1)
         ->  HashAggregate (actual time=600.452..624.715 rows=33422 loops=1)
               Group Key: "rollup".prelim_aggref_2, "rollup".a, "rollup"."grouping", "rollup"."group_id"
               Extra Text: (seg0)   33288 groups total in 32 batches; 1 overflows; 33288 spill groups.
 (seg0)   Hash chain length 4.3 avg, 14 max, using 10090 of 10240 buckets; total 2 expansions.
 
               ->  Subquery Scan on "rollup" (actual time=238.210..561.560 rows=33422 loops=1)
                     ->  Redistribute Motion 3:3  (slice1; segments: 3) (actual time=238.210..559.166 rows=33422 loops=1)
                           Hash Key: share0_ref1.b, share0_ref1.a, (Grouping), group_id()
                           ->  GroupAggregate (actual time=235.767..480.330 rows=33462 loops=1)
                                 Group Key: share0_ref1.b, share0_ref1.a
                                 ->  Sort (actual time=235.758..303.657 rows=167310 loops=1)
                                       Sort Key: share0_ref1.b, share0_ref1.a
                                       Sort Method:  external merge  Disk: 8192kB
                                       ->  Shared Scan (share slice:id 1:0) (actual time=53.192..61.242 rows=167310 loops=1)
                                             ->  Materialize (never executed)
                                                   ->  Seq Scan on test_src_tbl (actual time=0.020..30.024 rows=167310 loops=1)
         ->  HashAggregate (actual time=162.596..188.638 rows=33364 loops=1)
               Group Key: rollup_1.a, rollup_1.prelim_aggref_2, rollup_1."grouping", rollup_1."group_id"
               Extra Text: (seg0)   33281 groups total in 32 batches; 1 overflows; 33281 spill groups.
 (seg0)   Hash chain length 4.0 avg, 14 max, using 10020 of 10240 buckets; total 4 expansions.
 
               ->  Subquery Scan on rollup_1 (actual time=0.017..127.112 rows=33364 loops=1)
                     ->  Redistribute Motion 3:3  (slice2; segments: 3) (actual time=0.016..125.240 rows=33364 loops=1)
                           Hash Key: share0_ref2.a, share0_ref2.b, (Grouping), group_id()
                           ->  GroupAggregate (actual time=163.105..272.115 rows=33462 loops=1)
                                 Group Key: share0_ref2.a, share0_ref2.b
                                 ->  Sort (actual time=163.095..227.907 rows=167310 loops=1)
                                       Sort Key: share0_ref2.a, share0_ref2.b
                                       Sort Method:  external merge  Disk: 8192kB
                                       ->  Shared Scan (share slice:id 2:0) (actual time=0.024..12.089 rows=167310 loops=1)
```
### Conclusion
Based on the behaviors I observed, ORCA/planner on GPDB6 and ORCA on GPDB7 adapted a strategy based on Shared Scan and perform separated hashagg in top of two Shared Scan nodes. However, planner on GPDB7 supports "multiple hash tables in a hashagg node" and perform the hash works in a single node. The explain code reflects the work correctly.

According to my experiences, planner on GPDB7 is obviously faster than ORCA (924ms vs 1227ms) for grouping sets.

File: src/test/isolation2/init_file_isolation2 [hunk 7]
```diff
@@ -70,4 +70,45 @@ s/lsn: \d\/[0-9a-fA-F]+, prev \d\/[0-9a-fA-F]+/lsn: #\/########, prev #\/#######
...
+-- s/\d+K bytes wanted/###K bytes wanted/
+-- m/\d+ groups total in \d+ batches/
+-- s/\d+ groups total in \d+ batches/### groups total in ### batches/
+-- m/\d+ spill partitions/
+-- s/\d+ spill partitions/### spill partitions/
+-- m/disk usage: \d+/
+-- s/disk usage: \d+/disk usage: ###/
+-- m/using \d+ of \d+ buckets/
+-- s/using \d+ of \d+ buckets/using ## of ### buckets/
+-- m/\(actual time=\d+\.\d+..\d+\.\d+ rows=\d+ loops=\d+\)(\ +)/
```
Comment (adam8157) [inline src/test/isolation2/init_file_isolation2:91]: Why is this needed here and not needed in other cases?

**Comment by adam8157** [comment]
@yaowangm and I will pair resolving the left issues and then push this forward.

File: src/test/isolation2/init_file_isolation2 (hunk 7 above)
Comment (yaowangm) [reply src/test/isolation2/init_file_isolation2:91]: `init_file_isolation2`, like `src/test/regress/init_file`, is a common file which is included by all isolation2 test cases. So all isolation2 test cases are affected by the added code.

Comment (adam8157) [reply src/test/isolation2/init_file_isolation2:91]: I just noticed that the regression cases don't care about differences like `actual time`, `pg_regress` ignores issues like this:

```
<  Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..431.01 rows=20 width=12) (actual time=9.408..9.419 rows=20 loops=1)
<    ->  GroupAggregate  (cost=0.00..431.01 rows=7 width=12) (actual time=9.064..9.081 rows=7 loops=1)
>  Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..431.01 rows=20 width=12) (actual time=1.291..1.331 rows=20 loops=1)
>    ->  GroupAggregate  (cost=0.00..431.01 rows=7 width=12) (actual time=0.601..0.613 rows=7 loops=1)
```

**Comment by adam8157** [comment]
The dev CI triggered a deadlock likely fixed by the main branch already, rebase, and push forcibly again.

**Review by adam8157 — APPROVED** [review]
